Implementation elasticsearch dask parallel read.

avlahop opened the pull request:
- Edit _run_query to support slice scrolling
- Edit get_partition to support slice scrolling

Closes #6

avlahop commented:
@martindurant This is a first draft of the initial implementation. There are some issues that need addresing, and will update this PR. One of them and the most important is to send an empty dataframe but with the correct columns so this can aggree with the meta information provided in the `from_delayed` method. This is important since some shards might not return any results. These should be handled as an empty dataframe but with the correct columns and dtypes. 
Could you also provide with some small explanation on what is the role of the `self.part` attribute of the `ElasticsearchTableSource` object.

martindurant commented:
Thanks for posting @avlahop , I will look later today.

martindurant commented on intake_elasticsearch/elasticsearch_seq.py:
```diff
@@ -52,19 +52,27 @@ def __init__(self, query, qargs={}, metadata={}, **es_kwargs):
                                **self._qargs)
         except (JSONDecodeError, TypeError):
-            s = self.es.search(q=self._query, size=size, scroll=self._scroll,
+            s = self.es.search(body=slice_dict, q=self._query,
```
Can there be no slices if not using a JSON query?
You may not have gathered from the code, but the try: branch is the full query structure, which is written as a JSON string, and the except: branch is the simpler Lucene syntax.

avlahop replied:
I am not sure I get what you mean here. [Documentation](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-uri-request.html) lists the allowed parameters of the URI search (using `q`) and slice is not included.  The slice AFAIK needs to be passed in the payload of the `POST` operation of the search request. method.  That's why I am including the `slice_dict` as value of the `body` parameter.

martindurant replied:
I must admit I don't know the API well - but both partitioned and non-partitioned should be possible for both forms of the search method. It may well be that non-partitioned is exactly the same, but with the number of partitions equal to one, that might be the simplest way to code it.

avlahop replied:
I don't fully understand what you mean but that is the reason why I pass the `slice` payload so that both search methods support the slicing.
ES API is a bit vague although well documented (oxymoron) e.g if you want one partition passing the following payload won't work

```json
{"slice":{"id":0, "max":1}}
```

This would help for a much cleaner implementation IMHO.

martindurant commented on intake_elasticsearch/elasticsearch_seq.py:
```diff
@@ -85,12 +93,12 @@ def _get_schema(self, retry=2):
                            extra_metadata={})
 
-    def _get_partition(self, _):
+    def _get_partition(self, _, slice_id=None, slice_max=None):
```
The convention for the API of this function is `_get_partition(part)`, so in this case, `part` would be None in the case of no partitioning or (slice_id, slice_max) for parallel.

avlahop replied:
Ok I will pass it as a tuple of the two values.

martindurant commented on intake_elasticsearch/elasticsearch_table.py:
```diff
@@ -60,15 +59,24 @@ def _get_schema(self):
                            extra_metadata=self.metadata)
 
-    def to_dask(self):
+    def to_dask(self, npartitions=1):
```
npartitions should be in the source definition (i.e., argument to `__init__`), not here. That way the API here stays as in the superclass, without parameters.

avlahop replied:
Will do.

martindurant replied:
Yes, exactly. You could call the parameter something different, if you wish, but the source npartition should be updated in any case, this is something that users will want to know.

martindurant commented on intake_elasticsearch/elasticsearch_table.py:
```diff
@@ -60,15 +59,24 @@ def _get_schema(self):
+        else:
+            for slice_id in range(npartitions):
+                self._dataframe = None
+                part = delayed(self._get_partition(0, slice_id=slice_id,
```
should, I think, be delayed(self._get_partition)(args), otherwise, this is evaluated immediately, and only the result delayed, which is not useful. Indeed, I would do
```
dpart = delayed(self._get_partition)
for slice_id in range(npartitions):
    part = dpart(args)
    ...
```

avlahop replied:
I think you are right. Missed that.

martindurant commented on intake_elasticsearch/elasticsearch_table.py:
```diff
@@ -60,15 +59,24 @@ def _get_schema(self):
+        return dd.from_delayed(parts, meta=self.dtype)
 
-    def _get_partition(self, _):
+    def _get_partition(self, _, slice_id=None, slice_max=None):
```
See above.

martindurant commented on tests/test_intake_es.py:
```diff
@@ -113,6 +114,21 @@ def test_to_dask(engine):
     assert out[df.columns].equals(df)
 
 
+def test_to_dask_with_partitions(engine):
```
Please also add tests for a query which is JSON.

Do you plan for a partitioned sequence, which in dask is called a "bag", and would be constructed with `dask.bag.from_delayed`?

avlahop replied:
By JSON query you mean using the Query DSL right? i will proceed with adding one. 

If I understood correctly the `ElasticSearchSeqSource` ojbect is equivalent to a `JSON` response. Would you like me to add a `to_dask` method here?

martindurant replied:
Correct. The query can either be a string which can be interpreted as JSON (`"{...}"`) or a simple string like `'score:[0 TO 150]'`.

avlahop replied:
Take a look, added `to_dask` in ElasticSearchSeqSource` as well as JSON testcases.

avlahop commented:
Could you also take a look at the [build job](https://travis-ci.org/ContinuumIO/intake-elasticsearch/builds/447052623#L974)?

martindurant commented:
The job failure is not related, don't worry: the tests pass, but you don't have access to the anaconda token for uploading the dev package. That [line](https://github.com/ContinuumIO/intake-elasticsearch/blob/master/.travis.yml#L36) should probably be changed to only upload on a tagged version by removing the `else` branch.

martindurant commented on intake_elasticsearch/elasticsearch_seq.py:
```diff
@@ -77,12 +87,25 @@ def _run_query(self, size=None, end=None):
+                           npartitions=self.npartitions,
                            extra_metadata={})
 
     def _get_partition(self, _):
```
Would be nice to give the variable another name than `_`, now that it actually does something (and it should appear in the docstring, saying it can be a tuple).

avlahop replied:
Yes, to be honest, I didn't like it. You also call `_get_partition` with 0. In your `intake` documentation you mention that it's signature is `_get_partition(i)` where i is the number of partition, we need to get. Maybe something like:

```python
def _get_partition(partition=None):
    slice_id, slice_max = (partition, self.npartitions) if partition else (None, None)

    # rest of code here
```
`None` for partition would means get all
That way we keep the same signature as described in the `intake` documentation.

martindurant replied:
Thank sounds fine

avlahop replied:
Fixed you can check it. Have an issue with setting the tuple and using it in the `_run_query` so went with a slight different method.

martindurant commented on intake_elasticsearch/elasticsearch_seq.py:
```diff
@@ -92,5 +115,9 @@ def _get_partition(self, _):
+        slice_id = None
+        slice_max = None
+        if isinstance(_, tuple):
+            slice_id, slice_max = _
```
Combine the four lines as:
`slice_id, slice_max = _ if isinstance(_, tuple) else (None, None)` ?

martindurant commented on intake_elasticsearch/elasticsearch_table.py:
```diff
@@ -57,16 +56,25 @@ def _get_schema(self):
+        parts = []
+        if self.npartitions == 1:
+            part = delayed(self._get_partition(0))
+            parts.append(part)
```
`parts = [delayed(self._get_partition(0))]`; in practice, `to_dask()` is not useful when there's only one partition, so this may never get used.

avlahop replied:
Didn't know if you wanted previous functionality. Will remove it. Much clearer.

martindurant commented on intake_elasticsearch/elasticsearch_table.py:
```diff
@@ -76,13 +84,23 @@ def _get_partition(self, _):
-            results = self._run_query()
+            results = self._run_query(slice_id=slice_id, slice_max=slice_max)
             df = pd.DataFrame([r['_source'] for r in results['hits']['hits']])
             self._dataframe = df
```
Should not reset `self._dataframe` any more, I don't think, it should just remain the one derived from the first few rows of the dataset. The the following lines can look like
```
if df.empty:
    df = self._dataframe[:0]

..
return df
```

I am no longer sure what `self.part`'s purpose perhaps it's no longer needed.

avlahop commented:
I have re implemented the `to_dask` with one partition. Fixed typo, and also added two tests for the case of one partition.  Please check when you have the time. Will you resolve the discussions when you are  ready to merge?

martindurant commented:
Thank you! I'll clear it up to fix the CI after merge, and then release it.
